crates/wasi-common/src/snapshots/wasi_snapshot_preview1.rs

```diff
@@ -0,0 +1,1448 @@
+    ) -> Result<()> {
+        trace!(
+            "path_filestat_set_times(dirfd={:?}, flags={:?}, path={:?}, atim={}, mtim={}, fst_flags={})",
+            dirfd,
+            flags,
+            path,
+            atim,
+            mtim,
+            fst_flags,
+        );
+
+        let mut bc = GuestBorrows::new();
+        let as_str = path.as_raw(&mut bc)?;
+        let path = unsafe { &*as_str };
+
+        trace!("     | (path_ptr,path_len)='{}'", path);
+
+        let entry = unsafe { self.get_entry(dirfd)? };
+        let resolved = path::get(
+            &entry,
+            types::Rights::PATH_FILESTAT_SET_TIMES,
+            types::Rights::empty(),
+            flags,
+            path,
+            false,
+        )?;
+        match resolved.dirfd() {
+            Descriptor::VirtualFile(_virt) => {
+                unimplemented!("virtual filestat_set_times");
+            }
+            _ => path::filestat_set_times(resolved, flags, atim, mtim, fst_flags),
+        }
+    }
+
+    fn path_link(
+        &self,
+        old_fd: types::Fd,
+        old_flags: types::Lookupflags,
+        old_path: &GuestPtr<'_, str>,
+        new_fd: types::Fd,
+        new_path: &GuestPtr<'_, str>,
+    ) -> Result<()> {
+        trace!(
+            "path_link(old_fd={:?}, old_flags={:?}, old_path={:?}, new_fd={:?}, new_path={:?})",
+            old_fd,
+            old_flags,
+            old_path,
+            new_fd,
+            new_path,
+        );
+
+        // FIXME
+        // Much like in the case with `path_symlink`, we need to account for the
+        // fact where both `old_path` and `new_path` point the same memory location.
+        // Hence, separate scopes and `GuestBorrows` instances.
+        let resolved_old = {
+            let mut bc = GuestBorrows::new();
+            let old_as_str = old_path.as_raw(&mut bc)?;
+            let old_path = unsafe { &*old_as_str };
+
+            trace!("     | (old_path_ptr,old_path_len)='{}'", old_path);
+
+            let old_entry = unsafe { self.get_entry(old_fd)? };
+            path::get(
+                &old_entry,
+                types::Rights::PATH_LINK_SOURCE,
+                types::Rights::empty(),
+                types::Lookupflags::empty(),
+                old_path,
+                false,
+            )?
+        };
+        let resolved_new = {
+            let mut bc = GuestBorrows::new();
+            let new_as_str = new_path.as_raw(&mut bc)?;
+            let new_path = unsafe { &*new_as_str };
+
+            trace!("     | (new_path_ptr,new_path_len)='{}'", new_path);
+
+            let new_entry = unsafe { self.get_entry(new_fd)? };
+            path::get(
+                &new_entry,
+                types::Rights::PATH_LINK_TARGET,
+                types::Rights::empty(),
+                types::Lookupflags::empty(),
+                new_path,
+                false,
+            )?
+        };
+        path::link(resolved_old, resolved_new)
```

Comment:
I'll try to beat everyone to the question here. I'm not sure how to handle this better with the current state of `wiggle` (which is not to say we can't modify `wiggle` to accommodate this case). So, I was forced to create two instances of `GuestBorrows` since it is possible for the two `path`s that are handed to us from guest to point at the same memory location (one use case being for instance symlink loops, etc.).

I was thinking, we could potentially check for that by comparing the pointers (if they implemented `PartialEq`), but it still feels clunky. Perhaps we should revisit the concept of immutability in `wiggle`? That is, differentiate immutable vs mutable borrows? In this case, I reckon it should be fine to borrow the same location immutably multiple times. Thoughts? @alexcrichton what do you reckon?

Reply:
This is a great question. One alternative to creating two different borrowed path strings is to clone two path strings from the wasm memory into Rust owned `String`s. I expect, since paths shouldn't be too big, that is the right design choice here, except for the possibility of a DoS attack by creating two 4gb Strings here.

Reply:
Differentiating between mutable and immutable borrows sounds like a good way to go.

For example, POSIX doesn't say anything about [`link`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/link.html)'s arguments aliasing or not, but most functions with output buffers, such as [readlink](https://pubs.opengroup.org/onlinepubs/9699919799/functions/readlink.html) are defined with `restrict` qualifiers, meaning the output buffer is expected to not overlap with any of the input buffers. This also echos the Rust's borrow checker: you can have many aliasing immutable references, but a mutable reference must be unaliased.

Reply:
I've thought about this and I'm fairly convinced the right thing, for now, is to leave a nice comment that outlines the design decisions you just did in this thread.

Reply:
I don't actually think that `GuestBorrows` is really needed here at all because there are no long-lived pointers, I think it's fine to do what this is currently doing, but I might recommend making it a bit more concise with the comment I made above

Reply:
Hmm, I don't think I understand what you mean here by `GuestBorrows` not needed here at all. I'm probably missing something but let me outline the things I do believe:

I agree that in this case, the borrow checking is vacuous. But there isn't presently another method besides `GuestPtr::as_raw` that allows us to get a `&str` out of this `GuestPtr<str>`. We could add a method to GuestPtr that returns an owned `String`, but I fear that is a DoS vector because the guest will control the size of that String, and therefore gets to allocate in Rust's heap. We haven't yet rigorously audited this codebase for DoS vectors, of course, but I'm wary of adding something I know I'll have to forbid actually using in production.

I do not think we should create a shortcut variant of `GuestPtr::as_raw` to omit borrow checking which is only safe if the user ensures borrows are always short-lived, because I could see that leading to bugs when someone less familiar with the safety properties updates the code. Maybe that is too paranoid?

Reply:
Exactly as @pchickey explained, obtaining a reference to a slice-type object in Rust from `wiggle` is currently only possible through `as_raw` method which requires a `GuestBorrows` object in scope.

Reply:
@pchickey er sorry I don't think I was being very clear! I don't think we should remove the `GuestBorrows` argument or use owned strings here, I mean that each of these is resolved independently and there's never any overlapping borrows so documenting fixmes and/or hacks isn't necessary because each path is getting sucked into an owned object anyway so borrows never last long enough for runtime borrow-checking to be that relevant.